deeplake/integrations/mmdet/mmdet_.py

```diff
@@ -0,0 +1,1135 @@
+from collections import OrderedDict
+
+from typing import Callable, Optional, List, Dict
+from mmdet.apis.train import *
+from mmcv.utils import build_from_cfg
+from mmdet.datasets.builder import PIPELINES
+from mmdet.datasets.pipelines import Compose
+from mmcv.parallel import collate
+from functools import partial
+from deeplake.integrations.pytorch.dataset import TorchDataset
+from deeplake.client.client import DeepLakeBackendClient
+from mmdet.core import BitmapMasks
+import deeplake as dp
+from deeplake.util.warnings import always_warn
+import os.path as osp
+import warnings
+from collections import OrderedDict
+import mmcv
+from mmcv.runner import init_dist
+import torch
+import numpy as np
+from mmcv.utils import print_log
+from terminaltables import AsciiTable
+from mmdet.core import eval_map, eval_recalls
+from mmdet.datasets.pipelines import Compose
+from mmdet.utils.util_distribution import *
+import tempfile
+from deeplake.integrations.mmdet import mmdet_utils
+from deeplake.enterprise.dataloader import indra_available, dataloader
+from PIL import Image, ImageDraw
+import os
+
+
+class Dummy:
+    pass
+
+
+def force_cudnn_initialization(device_id):
+    dev = torch.device(f"cuda:{device_id}")
+    torch.nn.functional.conv2d(
+        torch.zeros(32, 32, 32, 32, device=dev), torch.zeros(32, 32, 32, 32, device=dev)
+    )
+
+
+def build_ddp(model, device, *args, **kwargs):
+    """Build DistributedDataParallel module by device type.
+
+    If device is cuda, return a MMDistributedDataParallel model;
+    if device is mlu, return a MLUDistributedDataParallel model.
+
+    Args:
+        model (:class:`nn.Module`): module to be parallelized.
+        device (str): device type, mlu or cuda.
+
+    Returns:
+        :class:`nn.Module`: the module to be parallelized
+
+    References:
+        .. [1] https://pytorch.org/docs/stable/generated/torch.nn.parallel.
+                     DistributedDataParallel.html
+    """
+
+    assert device in ["cuda", "mlu"], "Only available for cuda or mlu devices."
+    if device == "cuda":
+        model = model.cuda(kwargs["device_ids"][0])  # patch
+    elif device == "mlu":
+        from mmcv.device.mlu import MLUDistributedDataParallel
+
+        ddp_factory["mlu"] = MLUDistributedDataParallel
+        model = model.mlu()
+
+    return ddp_factory[device](model, *args, **kwargs)
+
+
+def coco_pixel_2_pascal_pixel(boxes, shape):
+    # Convert bounding boxes to Pascal VOC format and clip bounding boxes to make sure they have non-negative width and height
+
+    return np.stack(
+        (
+            boxes[:, 0],
+            boxes[:, 1],
+            boxes[:, 0] + boxes[:, 2],
+            boxes[:, 1] + boxes[:, 3],
+        ),
+        axis=1,
+    )
+
+
+def poly_2_mask(polygons, shape):
+    # TODO This doesnt fill the array inplace.
```

Comment:
?

Reply:
This is as fast as open cv without inplace fill, should be fine.